core/tee/tee_enc_fs.c

```diff
@@ -0,0 +1,450 @@
+
+	/* new created file? */
+	if (encrypted_data_len == 0) {
+		fd->len = 0;
+		fd->is_new_file = 1;
+		fd->data = NULL;
+		return handle_get(&fs_handle_db, fd);
+	}
+
+	/*
+	 * The file stored on normal world filesystem is encrypted data
+	 * with an additional file header
+	 */
+	TEE_ASSERT(encrypted_data_len > file_header_len);
+	file_len = encrypted_data_len - file_header_len;
+
+	/* allocate buffer to hold encrypted file context + header */
+	encrypted_data = malloc(encrypted_data_len);
+	if (!encrypted_data) {
+		res = -1;
+		goto exit_free_fd;
+	}
+
+	/*
+	 * init internal status, also allocate buffer for
+	 * decrypted file content
+	 */
+	fd->len = file_len;
+	fd->is_new_file = 0;
+	fd->data = malloc(file_len);
+	if (!fd->data) {
+		res = -1;
+		goto exit_free_encrypted_data;
+	}
+
+	/* read encrypted file content */
+	res = tee_fs_common_read(nw_fd, encrypted_data, encrypted_data_len);
+	if (res != (int)encrypted_data_len) {
+		res = -1;
+		goto exit_free_decrypted_data;
+	}
+	DMSG("%d bytes read", res);
+
+	/* decrypt and authenticate the file content */
+	ret = tee_enc_fs_file_decryption(encrypted_data, encrypted_data_len,
+			fd->data, &file_len);
+	if (ret != TEE_SUCCESS) {
+		res = -1;
+		goto exit_free_decrypted_data;
+	}
+
+	res = handle_get(&fs_handle_db, fd);
+
+exit_free_decrypted_data:
+	if (res < 0)
+		free(fd->data);
+exit_free_encrypted_data:
+	free(encrypted_data);
+exit_free_fd:
+	if (res < 0) {
+		if (fd->filename)
+			free(fd->filename);
+		free(fd);
+	}
+exit:
+	return res;
+}
+
+static int tee_enc_fs_close(int fd)
+{
+	int res = -1;
+	size_t encrypted_data_len;
+	size_t file_header_len = tee_enc_fs_get_file_header_size();
+	void *encrypted_data;
+	TEE_Result ret;
+	struct tee_enc_fs_fd *fdp = handle_put(&fs_handle_db, fd);
+
+	DMSG("close, fd=%d", fd);
+
+	if (!fdp)
+		return -1;
+
+	encrypted_data_len = fdp->len + file_header_len;
+
+	/* allocate buffer to hold encrypted file content */
+	encrypted_data = malloc(encrypted_data_len);
+	if (!encrypted_data)
+		goto exit_free_fd;
+
+	/* encrypt the file content and calculate MAC */
+	ret = tee_enc_fs_file_encryption(fdp->data, fdp->len,
+			encrypted_data, &encrypted_data_len);
+	if (ret != TEE_SUCCESS)
+		goto exit_free_encrypted_data;
+
+	res = tee_fs_common_lseek(fdp->nw_fd, 0, TEE_FS_SEEK_SET);
+	if (res < 0)
+		goto exit_free_encrypted_data;
+
+	res = tee_fs_common_ftruncate(fdp->nw_fd, encrypted_data_len);
+	if (res < 0)
+		goto exit_free_encrypted_data;
+
+	/* write encrypted file content to normal world file system */
+	res = tee_fs_common_write(fdp->nw_fd, encrypted_data,
+					encrypted_data_len);
+	if (res != (int)encrypted_data_len) {
+		res = -1;
+		goto exit_free_encrypted_data;
+	}
+	DMSG("%d bytes written", res);
+
+exit_free_encrypted_data:
+	free(encrypted_data);
+
+exit_free_fd:
+	tee_fs_common_close(fdp->nw_fd);
+
+	if (res < 0) {
+		EMSG("Fail to close file, start fail recovery function");
+		do_fail_recovery(fdp);
```

Comment:
In encrypted file system, we will try to encrypted data and write back to linux file system when close the file. 
If error occurred, we will start fail recovery mechanism and return error code. But most of functions 
ignore the return code when they call tee_file_ops.close(). So the TA will never know the file be deleted or be reverted to previous version by fail recovery mechanism. I think there have no easy way to fix the issue.